tui/network.py

```diff
@@ -9,128 +9,215 @@
-                    msg = 'Nameserver'
-            if vlan_cb.selected():
-                if not netutil.valid_vlan(vlan_field.value()):
-                    msg = 'VLAN'
-            if msg != '':
-                tui.progress.OKDialog("Networking", "Invalid %s, please check the field and try again." % msg)
+        if ipv6:
+            autoconf_rb = SingleRadioButton("Automatic configuration (Autoconf)", static_rb, 0)
+            autoconf_rb.setCallback(dhcp_change, ())
+        dhcp_change()
+
+        if defaults:
+            if ipv6:
+                if defaults.ipv6addr:
+                    ip6addr, netmask = defaults.ipv6addr.split("/")
+                    ip_field.set(ip6addr)
+                    subnet_field.set(netmask)
+                if defaults.ipv6_gateway:
+                    gateway_field.set(defaults.ipv6_gateway)
+            else:
+                if defaults.ipaddr:
+                    ip_field.set(defaults.ipaddr)
+                if defaults.netmask:
+                    subnet_field.set(defaults.netmask)
+                if defaults.gateway:
+                    gateway_field.set(defaults.gateway)
+
+            if defaults.dns:
+                dns_field.set(defaults.dns[0])
+
+        vlan_cb = Checkbox("Use VLAN:", defaults.isVlan() if defaults else False)
+        vlan_cb.setCallback(use_vlan_cb_change, ())
+        if defaults and defaults.isVlan():
+            vlan_field.set(str(defaults.vlan))
+        else:
+            vlan_field.setFlags(FLAG_DISABLED, False)
+
+        ip_msg = "IPv6 Address" if ipv6 else "IP Address"
+        mask_msg = "CIDR (4-128)" if ipv6 else "Subnet mask"
+        ip_text = Textbox(15, 1, "%s:" % ip_msg)
+        subnet_text = Textbox(15, 1, "%s:" % mask_msg)
+        gateway_text = Textbox(15, 1, "Gateway:")
+        dns_text = Textbox(15, 1, "Nameserver:")
+        vlan_text = Textbox(15, 1, "VLAN (1-4094):")
+
+        entry_grid = Grid(2, include_dns and 4 or 3)
+        entry_grid.setField(ip_text, 0, 0)
+        entry_grid.setField(ip_field, 1, 0)
+        entry_grid.setField(subnet_text, 0, 1)
+        entry_grid.setField(subnet_field, 1, 1)
+        entry_grid.setField(gateway_text, 0, 2)
+        entry_grid.setField(gateway_field, 1, 2)
+        if include_dns:
+            entry_grid.setField(dns_text, 0, 3)
+            entry_grid.setField(dns_field, 1, 3)
+
+        vlan_grid =  Grid(2, 1)
+        vlan_grid.setField(vlan_text, 0, 0)
+        vlan_grid.setField(vlan_field, 1, 0)
+
+        gf.add(text, 0, 0, padding=(0, 0, 0, 1))
+        gf.add(dhcp_rb, 0, 2, anchorLeft=True)
+        gf.add(static_rb, 0, 3, anchorLeft=True)
+        gf.add(entry_grid, 0, 4, padding=(0, 0, 0, 1))
+        if ipv6:
+            gf.add(autoconf_rb, 0, 5, anchorLeft=True)
+        # One more line for IPv6 autoconf
+        gf.add(vlan_cb, 0, 5 + ipv6, anchorLeft=True)
+        gf.add(vlan_grid, 0, 6 + ipv6, padding=(0, 0, 0, 1))
+        gf.add(buttons, 0, 7 + ipv6, growx=1)
```

Comment:
the common idiom in the TUI is to leave a blank line rather than complicating layout computations